official/coin_build/man/man3/SoNodeEngine.3
.TH "SoNodeEngine" 3 "Sun May 28 2017" "Version 4.0.0a" "Coin" \" -*- nroff -*-
.ad l
.nh
.SH NAME
SoNodeEngine \- \fBSoNodeEngine\fP is the base class for Coin node engines\&.
.PP
Node engines have the same functionality as normal engines, except that they inherit \fBSoNode\fP, which makes it possible to insert node engines in the scene graph\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Inventor/engines/SoNodeEngine\&.h>\fP
.PP
Inherits \fBSoNode\fP\&.
.PP
Inherited by \fBSoVRMLInterpolator\fP, and \fBSoVRMLTimeSensor\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "void \fBevaluateWrapper\fP (void)"
.br
.ti -1c
.RI "virtual \fBint\fP \fBgetOutputs\fP (\fBSoEngineOutputList\fP &l) const"
.br
.ti -1c
.RI "\fBSoEngineOutput\fP * \fBgetOutput\fP (const \fBSbName\fP &outputname) const"
.br
.ti -1c
.RI "SbBool \fBgetOutputName\fP (const \fBSoEngineOutput\fP *output, \fBSbName\fP &outputname) const"
.br
.ti -1c
.RI "virtual const \fBSoEngineOutputData\fP * \fBgetOutputData\fP (void) const =0"
.br
.ti -1c
.RI "SbBool \fBisNotifying\fP (void) const"
.br
.ti -1c
.RI "virtual void \fBnotify\fP (\fBSoNotList\fP *nl)"
.br
.ti -1c
.RI "virtual void \fBwriteInstance\fP (\fBSoOutput\fP *out)"
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static void \fBinitClass\fP (void)"
.br
.ti -1c
.RI "static \fBSoType\fP \fBgetClassTypeId\fP (void)"
.br
.in -1c
.SS "Protected Member Functions"

.in +1c
.ti -1c
.RI "\fBSoNodeEngine\fP (void)"
.br
.ti -1c
.RI "virtual \fB~SoNodeEngine\fP (void)"
.br
.ti -1c
.RI "virtual void \fBevaluate\fP (void)=0"
.br
.ti -1c
.RI "virtual SbBool \fBreadInstance\fP (\fBSoInput\fP *in, unsigned short flags)"
.br
.ti -1c
.RI "virtual void \fBinputChanged\fP (\fBSoField\fP *which)"
.br
.ti -1c
.RI "void \fBwriteOutputTypes\fP (\fBSoOutput\fP *out)"
.br
.in -1c
.SS "Static Protected Member Functions"

.in +1c
.ti -1c
.RI "static const \fBSoFieldData\fP ** \fBgetFieldDataPtr\fP (void)"
.br
.ti -1c
.RI "static const \fBSoEngineOutputData\fP ** \fBgetOutputDataPtr\fP (void)"
.br
.in -1c
.SS "Additional Inherited Members"
.SH "Detailed Description"
.PP 
\fBSoNodeEngine\fP is the base class for Coin node engines\&.
.PP
Node engines have the same functionality as normal engines, except that they inherit \fBSoNode\fP, which makes it possible to insert node engines in the scene graph\&. 

The main rationale for this class is to simplify the implementation of VRML interpolator nodes, which are in a sense engines embedded in the shape of ordinary nodes\&.
.PP
This abstract superclass will likely be of no interest to the Coin application programmer, and you can safely ignore it\&.
.PP
Be aware that this class is an extension for Coin, and it is not available in the original SGI Open Inventor v2\&.1 API\&. 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "SoNodeEngine::SoNodeEngine (void)\fC [protected]\fP"
Default constructor\&. 
.SS "SoNodeEngine::~SoNodeEngine (void)\fC [protected]\fP, \fC [virtual]\fP"
Destructor\&. 
.SH "Member Function Documentation"
.PP 
.SS "void SoNodeEngine::evaluateWrapper (void)"
Triggers an engine evaluation\&. 
.SS "\fBint\fP SoNodeEngine::getOutputs (\fBSoEngineOutputList\fP & l) const\fC [virtual]\fP"
Adds all outputs to list\&. Returns the number of outputs added to the list\&. 
.SS "\fBSoEngineOutput\fP * SoNodeEngine::getOutput (const \fBSbName\fP & outputname) const"
Returns the output with name \fIoutputname\fP, or \fCNULL\fP if no such output exists\&. 
.SS "SbBool SoNodeEngine::getOutputName (const \fBSoEngineOutput\fP * output, \fBSbName\fP & outputname) const"
Sets  to the name of \fIoutput\fP\&. Returns \fCFALSE\fP if no such output is contained within the engine instance\&. 
.SS "const \fBSoEngineOutputData\fP * SoNodeEngine::getOutputData (void) const\fC [pure virtual]\fP"
\fIThis API member is considered internal to the library, as it is not likely to be of interest to the application programmer\&.\fP 
.PP
Implemented in \fBSoVRMLTimeSensor\fP, \fBSoVRMLPositionInterpolator\fP, \fBSoVRMLOrientationInterpolator\fP, \fBSoVRMLColorInterpolator\fP, \fBSoVRMLCoordinateInterpolator\fP, \fBSoVRMLInterpolator\fP, \fBSoVRMLNormalInterpolator\fP, and \fBSoVRMLScalarInterpolator\fP\&.
.SS "SbBool SoNodeEngine::isNotifying (void) const"
Returns whether we're in a notification process\&. This is needed to avoid double notification when an engine enables outputs during \fBinputChanged()\fP\&. 
.SS "void SoNodeEngine::notify (\fBSoNotList\fP * l)\fC [virtual]\fP"
Notifies all auditors for this instance when changes are made\&. 
.PP
Reimplemented from \fBSoNode\fP\&.
.PP
Reimplemented in \fBSoVRMLTimeSensor\fP\&.
.SS "void SoNodeEngine::writeInstance (\fBSoOutput\fP * out)\fC [virtual]\fP"
Writes all the fields contained in this instance to the output stream within \fIout\fP\&.
.PP
This method is solely called from the write methods of fields\&.
.PP
Either from \fBSoField\fP if the write is done because of a field-to-field connection, or from one of the fields which may actually write \fBSoFieldContainer\fP instances, i\&.e\&. \fBSoSFNode\fP, \fBSoMFNode\fP, \fBSoSFEngine\fP, \fBSoMFEngine\fP, \fBSoSFPath\fP and \fBSoMFPath\fP\&.
.PP
This method, in addition to the ordinary \fBwrite()\fP method of \fBSoNode\fP, needs to be present since the fields don't have a write action instance in their writeValue() method, and \fBwriteInstance()\fP will create a new \fBSoWriteAction\fP and call continueToApply() on it\&. 
.PP
Reimplemented from \fBSoNode\fP\&.
.SS "void SoNodeEngine::evaluate (void)\fC [protected]\fP, \fC [pure virtual]\fP"
\fIThis API member is considered internal to the library, as it is not likely to be of interest to the application programmer\&.\fP 
.SS "SbBool SoNodeEngine::readInstance (\fBSoInput\fP * in, unsigned short flags)\fC [protected]\fP, \fC [virtual]\fP"
This method is mainly intended for internal use during file import operations\&.
.PP
It reads a definition of an instance from the input stream \fIin\fP\&. The input stream state points to the start of a serialized / persistant representation of an instance of this class type\&.
.PP
\fCTRUE\fP or \fCFALSE\fP is returned, depending on if the instantiation and configuration of the new object of this class type went ok or not\&. The import process should be robust and handle corrupted input streams by returning \fCFALSE\fP\&.
.PP
\fIflags\fP is used internally during binary import when reading user extension nodes, group nodes or engines\&. 
.PP
Reimplemented from \fBSoNode\fP\&.
.SS "void SoNodeEngine::inputChanged (\fBSoField\fP * which)\fC [protected]\fP, \fC [virtual]\fP"
Called when an input is changed\&. The default method does nothing, but subclasses may override this method to do the The Right Thing when a specific field is changed\&. 
.PP
Reimplemented in \fBSoVRMLTimeSensor\fP\&.
.SS "const \fBSoFieldData\fP ** SoNodeEngine::getFieldDataPtr (void)\fC [static]\fP, \fC [protected]\fP"
Returns the \fBSoFieldData\fP class which holds information about inputs in this engine\&. 
.SS "const \fBSoEngineOutputData\fP ** SoNodeEngine::getOutputDataPtr (void)\fC [static]\fP, \fC [protected]\fP"
Returns the \fBSoEngineOutputData\fP class which holds information about the outputs in this engine\&. 
.SS "void SoNodeEngine::writeOutputTypes (\fBSoOutput\fP * out)\fC [protected]\fP"
Writes the types of engine outputs for extension engines (i\&.e\&. engines not built in to Coin)\&. 

.SH "Author"
.PP 
Generated automatically by Doxygen for Coin from the source code\&.


coin4/src/engines/SoNodeEngine.cpp
/**************************************************************************\
 * Copyright (c) Kongsberg Oil & Gas Technologies AS
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 * 
 * Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 * 
 * Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 * 
 * Neither the name of the copyright holder nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
\**************************************************************************/

/*!
  \class SoNodeEngine SoNodeEngine.h Inventor/engines/SoNodeEngine.h
  \brief SoNodeEngine is the base class for Coin node engines.
  \ingroup engines

  Node engines have the same functionality as normal engines, except
  that they inherit SoNode, which makes it possible to insert node
  engines in the scene graph.

  The main rationale for this class is to simplify the implementation
  of VRML interpolator nodes, which are in a sense engines embedded in
  the shape of ordinary nodes.

  This abstract superclass will likely be of no interest to the Coin
  application programmer, and you can safely ignore it.

  \COIN_CLASS_EXTENSION
*/

// FIXME: currently most of the code in this class is simply copied
// from SoEngine.cpp. We should try to use a templant file or
// something to make it easier to maintain the code.
// pederb, 2001-10-23

#include <Inventor/engines/SoNodeEngine.h>
#include <Inventor/engines/SoEngineOutput.h>
#include <Inventor/engines/SoOutputData.h>
#include <Inventor/lists/SoEngineOutputList.h>
#include <Inventor/SbName.h>
#include <Inventor/errors/SoDebugError.h>

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif // HAVE_CONFIG_H
#include "coindefs.h" // COIN_STUB()
#ifdef COIN_THREADSAFE
#include "threads/recmutexp.h"
#endif // COIN_THREADSAFE

// *************************************************************************

// FIXME: document these properly. 20000405 mortene.
/*!
  \fn const SoEngineOutputData * SoNodeEngine::getOutputData(void) const
  \COININTERNAL
*/
/*!
  \fn void SoNodeEngine::evaluate(void)
  \COININTERNAL
*/

// *************************************************************************

SoType SoNodeEngine::classTypeId STATIC_SOTYPE_INIT;

#define FLAG_ISNOTIFYING 0x1

// *************************************************************************

/*!
  Default constructor.
*/
SoNodeEngine::SoNodeEngine(void)
{
  this->flags = 0;
}

/*!
  Destructor.
*/
SoNodeEngine::~SoNodeEngine()
{
#if COIN_DEBUG && 0 // debug
  SoDebugError::postInfo("SoNodeEngine::~SoNodeEngine", "%p", this);
#endif // debug
}

// Overrides SoBase::destroy().
void
SoNodeEngine::destroy(void)
{
#if COIN_DEBUG && 0 // debug
  SbName n = this->getName();
  SoType t = this->getTypeId();
  SoDebugError::postInfo("SoNodeEngine::destroy", "start -- '%s' (%s)",
                         n.getString(),
                         t.getName().getString());
#endif // debug

#ifdef COIN_THREADSAFE
  cc_recmutex_internal_field_lock();
#endif // COIN_THREADSAFE
  this->evaluateWrapper();
#ifdef COIN_THREADSAFE
  cc_recmutex_internal_field_unlock();
#endif // COIN_THREADSAFE

  inherited::destroy();

#if COIN_DEBUG && 0 // debug
  SoDebugError::postInfo("SoNodeEngine::destroy", "done -- '%s' (%s)",
                         n.getString(),
                         t.getName().getString());
#endif // debug
}

// Doc from superclass.
void
SoNodeEngine::initClass(void)
{
  SoNodeEngine::classTypeId =
    SoType::createType(SoNode::getClassTypeId(), SbName("NodeEngine"));
}

// Documented in superclass.
SoType
SoNodeEngine::getClassTypeId(void)
{
  return SoNodeEngine::classTypeId;
}

/*!
  Adds all outputs to list. Returns the number of outputs added to the
  list.
*/
int
SoNodeEngine::getOutputs(SoEngineOutputList & l) const
{
  const SoEngineOutputData * outputs = this->getOutputData();
  if (outputs == nullptr) return 0;

  int n = outputs->getNumOutputs();
  for (int i = 0; i < n; i++) {
    l.append(outputs->getOutput(this, i));
  }
  return n;
}

/*!
  Returns the output with name \a outputname, or \c nullptr if no such
  output exists.
*/
SoEngineOutput *
SoNodeEngine::getOutput(const SbName & outputname) const
{
  const SoEngineOutputData * outputs = this->getOutputData();
  if (outputs == nullptr) return nullptr;
  int n = outputs->getNumOutputs();
  for (int i = 0; i < n; i++) {
    if (outputs->getOutputName(i) == outputname)
      return outputs->getOutput(this, i);
  }
  return nullptr;
}

/*!
  Sets \outputname to the name of \a output. Returns \c false if no
  such output is contained within the engine instance.
*/
bool
SoNodeEngine::getOutputName(const SoEngineOutput * output,
                        SbName & outputname) const
{
  const SoEngineOutputData * outputs = this->getOutputData();
  if (outputs == nullptr) return false;

  int n = outputs->getNumOutputs();
  for (int i = 0; i < n; i++) {
    if (outputs->getOutput(this, i) == output) {
      outputname = outputs->getOutputName(i);
      return true;
    }
  }
  return false;
}

/*!
  Called when an input is changed. The default method does nothing,
  but subclasses may override this method to do the The Right Thing
  when a specific field is changed.
*/
void
SoNodeEngine::inputChanged(SoField * COIN_UNUSED_ARG(which))
{
}

// doc in parent
void
SoNodeEngine::notify(SoNotList * nl)
{
#if COIN_DEBUG && 0 // debug
  SoDebugError::postInfo("SoNodeEngine::notify", "%p - %s, start",
                         this, this->getTypeId().getName().getString());
#endif // debug

  // Avoid recursive notification calls.
  if (this->isNotifying()) return;

  this->flags |= FLAG_ISNOTIFYING;

  // Call inputChanged() only if we're being notified through one of
  // the engine's fields (lastrec == CONTAINER, set in
  // SoField::notify()).
  SoNotRec * lastrec = nl->getLastRec();
  if (lastrec && lastrec->getType() == SoNotRec::CONTAINER) {
    // Let engine know that a field changed, so we can recalculate
    // internal variables, if necessary.
    this->inputChanged(nl->getLastField());
  }

  // add ourself to the notification list
  SoNotRec rec(createNotRec());
  rec.setType(SoNotRec::ENGINE);
  nl->append(&rec);

  // Notify the slave fields connected to our engine outputs.
  const SoEngineOutputData * outputs = this->getOutputData();
  int numoutputs = outputs->getNumOutputs();
  for (int i = 0; i < numoutputs; i++)
    outputs->getOutput(this, i)->touchSlaves(nl, this->isNotifyEnabled());

  this->flags &= ~FLAG_ISNOTIFYING;

#if COIN_DEBUG && 0 // debug
  SoDebugError::postInfo("SoNodeEngine::notify", "%p - %s, done",
                         this, this->getTypeId().getName().getString());
#endif // debug
}

/*!
  Triggers an engine evaluation.
*/
void
SoNodeEngine::evaluateWrapper(void)
{
  const SoEngineOutputData * outputs = this->getOutputData();
  int i, n = outputs->getNumOutputs();
  for (i = 0; i < n; i++) {
    outputs->getOutput(this, i)->prepareToWrite();
  }
  this->evaluate();
  for (i = 0; i < n; i++) {
    outputs->getOutput(this, i)->doneWriting();
  }
}

/*!
  Returns the SoFieldData class which holds information about inputs
  in this engine.
*/
const SoFieldData **
SoNodeEngine::getFieldDataPtr(void)
{
  return nullptr; // base class has no input
}

/*!
  Returns the SoEngineOutputData class which holds information about
  the outputs in this engine.
*/
const SoEngineOutputData **
SoNodeEngine::getOutputDataPtr(void)
{
  return nullptr; // base class has no output
}

// Documented in superclass.
bool
SoNodeEngine::readInstance(SoInput * in, unsigned short flagsarg)
{
  // FIXME: I believe there's code missing here for reading
  // SoUnknownEngine instances. 20000919 mortene.
  return inherited::readInstance(in, flagsarg);
}

// Documented in superclass.
void
SoNodeEngine::writeInstance(SoOutput * out)
{
  inherited::writeInstance(out);
}

/*!
  Writes the types of engine outputs for extension engines
  (i.e. engines not built in to Coin).
*/
void
SoNodeEngine::writeOutputTypes(SoOutput * COIN_UNUSED_ARG(out))
{
  COIN_STUB();
}

/*!
  Returns whether we're in a notification process. This is needed to
  avoid double notification when an engine enables outputs during
  inputChanged().
*/
bool
SoNodeEngine::isNotifying(void) const
{
  return (this->flags & FLAG_ISNOTIFYING) != 0;
}

#undef FLAG_ISNOTIFYING
